ultravox/training/train.py

```diff
@@ -282,48 +297,103 @@ def main() -> None:
         ),
     )
-    if args.do_train:
-        # Training loop
-        logging.info("Starting training...")
-        t_start = datetime.now()
-        logging.info(f"train start time: {t_start}")
-        if args.val_steps:
-            trainer.evaluate()
-        trainer.train()
-        trainer.save_model(args.output_dir)
-        t_end = datetime.now()
-        logging.info(f"train end time: {t_end}")
-        logging.info(f"elapsed: {t_end - t_start}")
-
-    if args.do_eval:
-        logging.info("Starting evaluation...")
-        t_start = datetime.now()
-        logging.info(f"eval start time: {t_start}")
-
-        # Merge LoRA weights for better inference performance.
-        # Note: this is irreversible and changes model saving format
-        model.merge_and_unload()
-        inference = infer.LocalInference(
-            model=model,
-            processor=processor,
-            tokenizer=text_tokenizer,
-            device=args.device,
-            dtype=dtype,
+
+    # Training loop
+    logging.info("Starting training...")
+    t_start = datetime.now()
+    logging.info(f"train start time: {t_start}")
+    if args.val_steps:
+        trainer.evaluate()
+    trainer.train()
+    if trainer.is_world_process_zero():
+        # Saving the model using pipeline to ensure its code is saved
+        pipeline = ultravox_pipeline.UltravoxPipeline(
+            model, tokenizer=text_tokenizer, device=device
         )
-        metrics = evaluation.evaluate(
-            inference,
-            data_dir=args.data_dir,
-            num_procs=args.eval_num_procs,
-            num_samples=args.eval_num_samples,
-            max_new_tokens=args.eval_max_new_tokens,
-            log_dir=wandb.run.dir if wandb.run else str(args.logs_dir),
+        pipeline.save_pretrained(args.output_dir)
+    t_end = datetime.now()
+    logging.info(f"train end time: {t_end}")
+    logging.info(f"elapsed: {t_end - t_start}")
+
+
+def evaluate(args: config_base.TrainConfig):
+    """
+    Evaluate the model on the audio and text datasets.
+
+    NOTE: This function must be run only on the primary process.
+    """
+    logging.info("Starting evaluation...")
+    t_start = datetime.now()
+    logging.info(f"eval start time: {t_start}")
+
+    if args.text_model_lora_config and args.text_model_lora_config.r:
+        logging.warn(
+            "Model has unmerged LoRA config. This can lead to slower inference."
         )
-        if is_master:
-            trainer.log(metrics)
 
-        t_end = datetime.now()
-        logging.info(f"eval end time: {t_end}")
-        logging.info(f"elapsed: {t_end - t_start}")
+    logs_dir = wandb.run.dir if wandb.run else str(args.logs_dir)
+
+    # Run audio-based evaluations and log to W&B
+    audio_metrics = run_oaievalset(
+        log_dir=os.path.join(logs_dir, "oaieval/audio"),
+        model_dir=str(args.output_dir),
+        eval_set="audio-core",
+        num_samples=args.eval_num_samples,
```

Comment:
Ideally, `eval_num_samples` should be configurable for each eval dataset separately. We can address this in another PR.

Reply:
Yeah originally I wanted to run the full set even, but with the multiple GPU issue I ended up restricting it again so that runs wouldn't take too long (though the default num_samples has already been increased by 4x). Hence I view this as a more of a temporary thing.